codec-http/src/main/java/io/netty/handler/codec/http/DefaultHttpHeaders.java

```diff
@@ -15,60 +15,109 @@
+import io.netty.util.internal.PlatformDependent;
 
+import java.util.AbstractList;
 import java.util.Calendar;
-import java.util.Comparator;
 import java.util.Date;
+import java.util.Iterator;
 import java.util.List;
-import java.util.TreeMap;
+import java.util.Map.Entry;
 
+import static io.netty.util.AsciiString.CASE_INSENSITIVE_HASHER;
+import static io.netty.util.AsciiString.CASE_SENSITIVE_HASHER;
 import static io.netty.util.internal.ObjectUtil.checkNotNull;
 
-public class DefaultHttpHeaders extends DefaultTextHeaders implements HttpHeaders {
-
-    private static final int HIGHEST_INVALID_NAME_CHAR_MASK = ~63;
+public class DefaultHttpHeaders extends DefaultHeaders<CharSequence> implements HttpHeaders {
     private static final int HIGHEST_INVALID_VALUE_CHAR_MASK = ~15;
+    private static final ByteProcessor HEADER_NAME_VALIDATOR = new ByteProcessor() {
+        @Override
+        public boolean process(byte value) throws Exception {
+            switch (value) {
+            case '\t':
+            case '\n':
+            case 0x0b:
+            case '\f':
+            case '\r':
+            case ' ':
+            case ',':
+            case ':':
+            case ';':
+            case '=':
+                throw new IllegalArgumentException(
+                      "a header name cannot contain the following prohibited characters: =,;: \\t\\r\\n\\v\\f: " +
+                          value);
+            default:
+                if (value < 0) {
+                    throw new IllegalArgumentException("a header name cannot contain non-ASCII character: " +
+                          value);
+                }
+            }
+            return true;
+        }
+    };
+    static final NameValidator<CharSequence> HttpNameValidator = new NameValidator<CharSequence>() {
+        @Override
+        public void validateName(CharSequence name) {
+            if (name instanceof AsciiString) {
+                try {
+                    ((AsciiString) name).forEachByte(HEADER_NAME_VALIDATOR);
+                } catch (Exception e) {
+                    PlatformDependent.throwException(e);
+                }
+            } else {
+                checkNotNull(name, "name");
+                // Go through each character in the name
+                for (int index = 0; index < name.length(); ++index) {
+                    char character = name.charAt(index);
+                    switch (character) {
+                    case '\t':
+                    case '\n':
+                    case 0x0b:
+                    case '\f':
+                    case '\r':
+                    case ' ':
+                    case ',':
+                    case ':':
+                    case ';':
+                    case '=':
+                        throw new IllegalArgumentException(
+                           "a header name cannot contain the following prohibited characters: =,;: \\t\\r\\n\\v\\f: " +
+                                    name);
```

Comment:
I can do this. During performance tests I didn't want to have to go through all the casting and masking stuff.